### PR TITLE
Add builder classes, make them the only parameter to macro methods

### PR DESCRIPTION
**Pull request by jakemac53**
This is still a WIP, but I wanted to get it out there for an initial review.

- Static analysis now is clean
- JsonCodable works, haven't tried the other ones
- I still need to get all the tests passing (I am sure many fail, haven't actually checked), will address this tomorrow
- All builder classes still have the same `query` API, which needs to be evaluated and hopefully restricted per phase in some way.
- The QueryClass macro was actually breaking the rules. It is no longer allowed to, so I changed a bit how it works.

### CODE REVIEW

**Review by davidmorgan — APPROVED** [review]
Looks good, thanks.

File: pkgs/macro/lib/src/builders.dart [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+
+import 'macro.dart';
+
+/// The base interface used to add declarations to the program as well
+/// as augment existing ones.
+abstract interface class Builder {
+  /// The target declaration of this macro application.
+  ///
+  /// All builder subtypes should override this with a more specific type.
+  Object get target;
```
[the remark this answers is not in the document]
Comment (jakemac53) [reply pkgs/macro/lib/src/builders.dart:15]: Sure, I think that makes sense.

**Comment by jakemac53** [comment]
@davidmorgan I made some substantive changes to fix some of the tests that you should look at - lazy merged maps are now considered equal if their left/right maps are equal, and `_keyof` and `getParents` also were modified for lazy merged maps.

**Comment by jakemac53** [comment]
To fix the remaining tests I had to change the expectations, the `qualifiedNameOf` function now technically works for cycles and re-used nodes (although the result could be surprising especially in the re-used nodes case).

I am not sure if this is a bad thing or a good thing... but these nodes shouldn't be re-used anyways.

**Comment by davidmorgan** [comment]
Thanks! Still LGTM.

Re: lists and cycles, both seem fine for now.

**Review by davidmorgan — APPROVED** [review]

**Pull request merged**